lib/join/ec2join/ec2.go

```diff
@@ -0,0 +1,409 @@
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func desktopServiceExists(ctx context.Context, presence services.Presence, hostID string) (bool, error) {
+	svcs, err := presence.GetWindowsDesktopServices(ctx)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	for _, wds := range svcs {
+		if wds.GetName() == hostID {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func resourceExists(ctx context.Context, presence services.Presence, role types.SystemRole, hostID string) (bool, error) {
+	switch role {
+	case types.RoleNode:
+		return nodeExists(ctx, presence, hostID)
+	case types.RoleProxy:
+		return proxyExists(ctx, presence, hostID)
+	case types.RoleKube:
+		return kubeExists(ctx, presence, hostID)
+	case types.RoleApp:
+		return appExists(ctx, presence, hostID)
+	case types.RoleDatabase:
+		return dbExists(ctx, presence, hostID)
+	case types.RoleWindowsDesktop:
+		return desktopServiceExists(ctx, presence, hostID)
+	case types.RoleOkta:
+		return oktaExists(ctx, presence, hostID)
+	case types.RoleDiscovery, types.RoleMDM:
+		// No appropriate check exists for these roles.
+		return false, nil
+	default:
+		return false, trace.BadParameter("role %s is unsupported for the EC2 join method", role)
+	}
+}
+
+func roleHasPresenceCheck(role types.SystemRole) bool {
+	switch role {
+	case types.RoleNode, types.RoleProxy, types.RoleKube, types.RoleApp,
+		types.RoleDatabase, types.RoleWindowsDesktop, types.RoleOkta:
+		return true
+	}
+	return false
+}
+
+// tryToDetectIdentityReuse performs a best-effort check to see if the specified role+id combination
+// is already in use by an instance. This will only detect re-use in the case where a recent heartbeat
+// clearly shows the combination in use since teleport maintains no long-term per-instance state.
+func tryToDetectIdentityReuse(ctx context.Context, params *CheckEC2RequestParams, hostID string) error {
+	if params.Role == types.RoleInstance {
+		// There is no "Instance" heartbeat, instead this must check if there
+		// are any heartbeats for any of the system roles granted by this
+		// token.
+		if params.SkipInstanceReuseCheck {
+			// This check must be skipped if called via the legacy join
+			// endpoint (where the Instance role may join after another sytem
+			// role may have joined and hearbeated) or if this is an
+			// authenticated re-join request (where existing heartbeats are
+			// expected and re-joins are explicitly allowed).
+			return nil
+		}
+		for _, role := range params.ProvisionToken.GetRoles() {
+			if !roleHasPresenceCheck(role) {
+				continue
+			}
+			alreadyExists, err := resourceExists(ctx, params.Presence, role, hostID)
+			if err != nil {
+				return trace.Wrap(err, "checking if %s with ID %s already exists", params.Role, hostID)
+			}
+			if alreadyExists {
+				return trace.AccessDenied("server with host ID %q and role %q already exists", hostID, role)
+			}
+		}
+		return nil
+	}
```

Comment:
This is new for the new join service. `Instance` is now the only role that anything actually joins with, hosts then use the instance certs to request other certs. So I added the presence check for all system roles granted by the token (which is the set of roles that will be included in the Instance cert)